Automattic/woocommerce-payments
# Add support for Japan to the tax details modal

Fixes #8926

#### Changes proposed in this Pull Request
This PR changes the tax details modal so it works well for merchants in Japan, alongside existing VAT merchants in EU and UK.

Prior to this PR, the modal assumes `VAT` and various other details of the flow are tailored to the currently-supported tax regions:

- Use of "VAT" throughout
- Specific hints for formatting of the VAT number
- Business address details are powered by the validation API (at least for some regions, e.g. UK)

To make the modal work for Japan, I've generalised it a little, by parameterising all the data/copy/content that is region-specific:

- The name of the tax ID is factored out (e.g. `VAT number` vs `Corporate Number`; in future `GST number`).
- The name of the tax is factored out (e.g. `VAT` vs `Consumption Tax`; in future `GST`).
- The validation hint is customisable per region.

I've made some other changes so things are clear and accurate

- Generalised the title, and made it more friendly/clear - now `Set your tax details` (covers all tasks - id and address).
- Fixed mention of `receipts` to `tax documents`, more consistent with rest of the UI and documentation.

##### here's what's changed, in visual form
In the first screenshot, I've highlighted the non-essential stuff I've changed.

| Step | Old - UK | New - UK | New – Japan |
| --- | --- | --- | --- |
| Open modal | <img width="764" alt="Existing UK 1" src="https://github.com/user-attachments/assets/98df3bf4-0b39-411b-a487-6cb7b0e365d5"> | <img width="729" alt="New UK 1" src="https://github.com/user-attachments/assets/97e123db-3737-4577-873f-395e79421542"> | <img width="680" alt="JP 1" src="https://github.com/user-attachments/assets/96cebb66-138a-48c7-b5fc-e9a83a6ac95c"> |
| Bad tax ID † | <img width="738" alt="Existing UK 2 - invalid VAT ID" src="https://github.com/user-attachments/assets/4f0b1fab-772b-4fd8-a207-1850b8e11a94"> | <img width="728" alt="New UK 2 - invalid" src="https://github.com/user-attachments/assets/126feef2-08d6-4267-b2c4-ee69f2f9c846"> | <img width="676" alt="JP 2 - invalid ID" src="https://github.com/user-attachments/assets/29cf1db1-ffa7-4015-b729-932218a90c82"> |
| Confirm details | <img width="725" alt="Existing UK 3 - enter details, valid ID" src="https://github.com/user-attachments/assets/67cadeb8-6598-44f2-ab44-5137afb773b1"> | <img width="726" alt="New UK 3 - details" src="https://github.com/user-attachments/assets/775d1baa-bf91-4619-9a78-d0e13362fb86"> | <img width="677" alt="JP 3 - valid, enter details" src="https://github.com/user-attachments/assets/06bd3ec1-5d06-4bcf-95ba-a6223ff9c7ce"> |

@rogermattic keen for your feedback on this, if you see anything amiss please comment!

† Note: The error message for a bad corporate number / Japan tax ID is currently incorrect. I've made server changes to allow us to customise these error messages, I'm proposing doing that in a follow up PR to keep this PR moving. 

- https://github.com/Automattic/woocommerce-payments/issues/9144


##### this is arguably a temporary fix
IMO this dialog is quite convoluted and complex, since each region is subtly different.

Long term, I think we should implement separate modals/components for each region, with a similar interface, and have one outer conditional for region. I didn't undertake that on this PR because I wanted a relatively quick and low-risk fix. 

Why: If each region's UI is self-contained, it's easier to review and test, or customise the flow if needed. Also IMO it's not good practice to translate substrings such as `VAT number`, it's better to translate whole sentences (or ideally whole modal/UI).

Next time we add a new region, or do any work, let's consider that approach. _Reviewers:_ keen for feedback on this aspect specifically.

#### Testing instructions
Note you'll need to test this with a store with [account country that supports tax docs](https://github.com/Automattic/woocommerce-payments-server/blob/8f6bdc2195cd0a99abf888300b292795e7d294c7/server/wp-content/rest-api-plugins/endpoints/wcpay/utils/class-vat-utils.php#L19). Ideally, test with different accounts to see UI for different countries. Tip: if you set `wcpay_accounts[].deleted` to a date in the past, you can onboard a new account. Or switch between accounts by setting wcpay_accounts[].deleted=NULL` on the account you want to use.

1. Set up tax docs so you have a document to view at `Payments > Documents`. Fake one up with this CLI command:
   - `wp wcpay add_document_for_account [account_id] vat_invoice --mode=test --period_from="2024-04-01 00:00:00" --period_to="2024-04-30 23:59:59"`
   - Or can use `wp wcpay backfill_merchant_vat_invoices`
1. If you've added VAT id / tax information previously, clear it out so you can see the modal: 
   - Delete `company_vat_submitted` `company_vat_number` `company_name` `company_address` from  `wcpay_account_meta` table.
   - Side note:
       - https://github.com/Automattic/woocommerce-payments-server/issues/5915
1. Go to `Payments > Documents`. 
1. Click `Download` for a tax doc row.
1. See modal.
1. Test adding valid and invalid tax info.
1. Test downloading the generated tax doc – confirm the merchant details are correct.


-------------------


**Post merge**



- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _not requesting GlobalStep testing, will need Japan store onboarded to test with, better to just test ourselves_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. _not affecting a critical flow; tax docs is a secondary feature_
